detectindent.lua

```diff
@@ -5,19 +5,50 @@ local util = import("micro/util")
 
 function onBufferOpen(buf)
     local spaces, tabs = 0, 0
+    local space_count, prev_space_count = -1, -1
+    local tabsizes = {}
     local i = 0
     while spaces + tabs < 500 and i < 1000 and i < buf:LinesNum() do
-        local r = util.RuneAt(buf:Line(i), 0)
+        space_count = -1
+        local line = buf:Line(i)
+        local r = util.RuneAt(line, 0)
         if r == " " then
             spaces = spaces + 1
+            space_count = string.len(util.GetLeadingWhitespace(line))
+            -- count empty lines as having the same indentation as the last one
+            if string.len(line) == space_count then
+                space_count = prev_space_count
+            end
         elseif r == "\t" then
             tabs = tabs + 1
         end
+        -- count the change in indentation between non-empty indented lines
+        if space_count > 0 and prev_space_count > 0 and space_count > prev_space_count then
```

Comment:
I see you ignore indented lines following non-indented lines, so e.g. tabsize is not detected with a simple file like this:

```
int main(void)
{
  return 0;
}
```

I actually like it because I see it useful in some other, more important cases (I can elaborate on that separately), so I'm gonna merge your PR anyway. Just curious: did you have any specific reasons to add this `prev_space_count > 0` check?

Reply:
Ah good catch, thanks! I've fixed that and simplified the logic. The reason to still check that `prev_space_count >= 0` is to catch the condition where it's -1, for example at the first line of the buffer or after a line indented with tabs.

Reply:
I am curious why it would be a feature to not detect the indent size for the example program you gave, since my failure to handle that was unintentional. But if that's the behavior you want, the condition would just be `prev_space_count > 0` instead of `prev_space_count >= 0`.

Reply:
Ok, first, I've pulled your fix and I see it still doesn't detect tabsize for my example. :) It seems `space_count = -1` should be `space_count = 0` at line 12.

Regarding usefulness of this behavior: consider a C file like this:

```
/*
 * This
 * is
 * a long
 * comment.
 */

int main(void)
{
	return 0;
}
```

The plugin currently misdetects it and sets tabstospaces=on even though the code in main() is indented with tabs.
This is a real-life issue which I encounter quite often. There are quite a few such files (with more multi-line comments than code) in Linux kernel source tree, for instance.
With your unintentional bug, at least it doesn't set tabsize to 1 (I remember it was a major annoyance for me with Vim detectindent plugin).
So your PR inspired me with a possible trick to solve this issue: for both tabstospaces and tabsize, ignore indented lines following non-indented lines.

Reply:
I see! I'll work on a fix that covers those cases and add commentary to explain.

Reply:
Okay I've solved it, but let me know if the method used is too dirty or if there's a cleaner way. It does handle your C example with a long block comment perfectly though.

Reply:
Unfortunately it seems there is no cleaner way, due to a limitation of [gopher-luar](https://pkg.go.dev/layeh.com/gopher-luar) which is used by Micro for Go->Lua mapping. In Go numbers and other primitive types can have methods (so Group has String()) but gopher-luar seems to support calling methods from Lua for non-primitive types only.

But that's not the main problem. I've found that your isCommentLine solution works only with very small files like my toy C example. For bigger files like [this one](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/plain/include/linux/uaccess.h?h=v5.13-rc3) it doesn't work most of the time. The reason is that syntax highlighting is initialized in the background after a file has been opened, when onBufferOpen is already running. We have a race condition here.

Nevertheless, thanks for trying things out.

BTW one more point: even if without this race condition it still would be not 100% perfect: it doesn't work if syntax highlighting is disabled.

Reply:
Is there a way to reference `comment.lua` since it's a built-in plugin now? I considered just duplicating its code but obviously it would be preferable to reuse it. Although I don't think it actually parses multiline comments, in which case it would need to be extended to fix your example. Let me know how to continue from here and I can hack on it some more next week.

Reply:
Well another option is to just discard indents of 1. Looks like VSCode took that approach as well as originally limiting it to even numbers as you can see in the response to this issue: https://github.com/microsoft/vscode/issues/62712

Reply:
The comment plugin indeed doesn't handle multiline comments.

I like the idea of just discarding indents of 1, for its simplicity.